### DISCUSSION TITLE
Modifying the Syntax Tree

### DISCUSSION THREAD

**Discussion by p-foucht** [tree-sitter/tree-sitter#2667] [Q&A] [answered]
Hey folks

I'd like to traverse a parsed tree while making updates to nodes (sometimes even removing/adding new nodes). Is there any support for this, or is there a recommended way to achieve this?

Thanks for any help!

**Comment by ahlinc** [comment]
Could you explain with a real example why you want to modify nodes?

Tree-sitter uses an another approach by allowing inserting or removing pieces of text and re-parse an edited text much faster within an old tree. You can think about it as an ability to edit trees by providing a text snippets of changes.

**Reply by p-foucht** [reply]
Sure!

I have tree-sitter set up for SQL (Presto SQL, specifically). Here is what I'm wanting to build.

Imagine the user has the following SQL in an editor.

`SELECT col1, col2, FROM table`

The trailing comma after col2 (`col2,`) is a syntax error. In response to some event, I'd like to parse the SQL, remove the trailing comma node.

`SELECT col1, col2, FROM table` --> `SELECT col1, col2 FROM table`

- Parse the SQL
- Walk the tree and identify the error comma node
- Remove the comma node (or really, make any modification to a node)
- Convert to the original string, with the comma removed

I'd like to retain all the previous formatting (newlines, spacing, etc) of the original text. I was wondering if there was any related functionality to accomplish this or if this was something I'd need to build out myself?

Thanks for responding @ahlinc!

**Reply by p-foucht** [reply]
@ahlinc Any suggestions by chance?

**Reply by ahlinc** [reply, marked as answer]
_Sorry, I've missed previous notification._

You need to built it by yourself on top of Tree-sitter, you can see more info in a similar discussion: https://github.com/tree-sitter/tree-sitter/discussions/2553#discussioncomment-7296041

**Reply by p-foucht** [reply]
Thanks for getting back to me, that answers my question. Thanks!

**Reply by ahlinc** [reply]
@p-foucht you can also look an interesting talk where also an another project author mentions how they use Tree-sitter for edits https://www.youtube.com/watch?v=NcUJnmBqHTY

Also by having positions for tokens nothing stops you to calculate positions for spaces between tokens to duplicate interested tokens and spaces into a resulting text by dropping some parts or inserting or modifing another parts. All this may be done while traversing an original tree.
